validator/src/machine/keygen/confirmed.ts

```diff
@@ -25,56 +31,105 @@ export const handleKeyGenConfirmed = async (
 	}
 	const groupId = event.gid;
 
-	// Check if this is our own confirmation
-	const ourParticipantId = keyGenClient.participantId(groupId);
-	if (event.identifier !== ourParticipantId) {
-		// Not our confirmation, just track
-		logger?.(`Group ${event.gid} confirmation from ${event.identifier} (not ours)`);
+	// Track this confirmation
+	const confirmedParticipants = [...machineStates.rollover.confirmedParticipants, event.identifier];
+	const participants = signingClient.participants(groupId);
+	const allConfirmed = confirmedParticipants.length === participants.length;
+
+	logger?.(
+		`Group ${groupId} confirmation from ${event.identifier} (${confirmedParticipants.length}/${participants.length})`,
+	);
+
+	// Genesis group: after all confirmations, we're done with keygen
+	if (consensusState.genesisGroupId === groupId) {
+		if (!allConfirmed) {
+			// Still waiting for confirmations
+			return {
+				rollover: {
+					...machineStates.rollover,
+					confirmedParticipants,
+					lastParticipant: event.identifier,
+				},
+			};
+		}
+		// All confirmed for genesis group - start preprocessing and return to waiting state
+		logger?.("Genesis group all confirmations received, starting preprocessing");
+		const consensus: ConsensusDiff = {
+			groupPendingNonces: [groupId, true],
+		};
+		const nonceTreeRoot = signingClient.generateNonceTree(groupId);
```

Comment:
As we do this in the epoch staged event I would not do this here, it feels duplicated